pkg/model/model.go

```diff
@@ -0,0 +1,295 @@
+			headNames = append(headNames, head.Name)
+		}
+		return nil, fmt.Errorf("multiple channel heads found in graph: %s", strings.Join(headNames, ", "))
+	}
+	return heads[0], nil
+}
+
+func (c *Channel) Validate() error {
+	var result *multierror.Error
+	if c.Name == "" {
+		result = multierror.Append(result, errors.New("channel name must not be empty"))
+	}
+
+	if c.Package == nil {
+		result = multierror.Append(result, errors.New("package must be set"))
+	}
+
+	if len(c.Bundles) == 0 {
+		result = multierror.Append(result, fmt.Errorf("channel must contain at least one bundle"))
+	}
+
+	if len(c.Bundles) > 0 {
+		if _, err := c.Head(); err != nil {
+			result = multierror.Append(result, err)
+		}
+	}
+
+	for name, b := range c.Bundles {
+		if name != b.Name {
+			result = multierror.Append(result, fmt.Errorf("bundle key %q does not match bundle name %q", name, b.Name))
+		}
+		if err := b.Validate(); err != nil {
+			result = multierror.Append(result, fmt.Errorf("invalid bundle %q: %v", b.Name, err))
+		}
+		if b.Channel != c {
+			result = multierror.Append(result, fmt.Errorf("bundle %q not correctly linked to parent channel", b.Name))
+		}
+	}
+	return result.ErrorOrNil()
+}
+
+type Bundle struct {
+	Package       *Package
+	Channel       *Channel
+	Name          string
+	Image         string
+	Replaces      string
+	Skips         []string
+	Properties    []property.Property
+	RelatedImages []RelatedImage
+
+	// These fields are present so that we can continue serving
+	// the GRPC API the way packageserver expects us to in a
+	// backwards-compatible way.
+	Objects []string
+	CsvJSON string
+}
+
+func (b *Bundle) Validate() error {
+	var result *multierror.Error
+	if b.Name == "" {
+		result = multierror.Append(result, errors.New("name must be set"))
+	}
+	if b.Channel == nil {
+		result = multierror.Append(result, errors.New("channel must be set"))
+	}
+	if b.Package == nil {
+		result = multierror.Append(result, errors.New("package must be set"))
+	}
+	if b.Channel != nil && b.Package != nil && b.Package != b.Channel.Package {
+		result = multierror.Append(result, errors.New("package does not match channel's package"))
+	}
+	if b.Replaces != "" {
+		if b.Channel != nil && b.Channel.Bundles != nil {
+			if _, ok := b.Channel.Bundles[b.Replaces]; !ok {
+				result = multierror.Append(result, fmt.Errorf("replaces %q not found in channel", b.Replaces))
+			}
+		}
+	}
+	props, err := property.Parse(b.Properties)
+	if err != nil {
+		result = multierror.Append(result, err)
+	}
+	for i, skip := range b.Skips {
+		if skip == "" {
+			result = multierror.Append(result, fmt.Errorf("skip[%d] is empty", i))
+		}
+	}
+	// TODO(joelanford): Validate related images? It looks like some
+	//   CSVs in production databases use incorrect fields ([name,value]
+	//   instead of [name,image]), which results in empty image values.
+	//   Example is in redhat-operators: 3scale-operator.v0.5.5
+	//for i, relatedImage := range b.RelatedImages {
```

Comment:
Possibly? But related images is a first-class field with a well-defined API. If anything, perhaps this check is redundant because other components are catching it, but clearly that isn't the case because existing CSVs in production catalogs were failing this until I commented it out.